lib/vector-core/src/transform/mod.rs

```diff
@@ -13,6 +13,7 @@ mod config;
 /// transforms act as a coordination or barrier point.
 pub enum Transform {
     Function(Box<dyn FunctionTransform>),
+    FallibleFunction(Box<dyn FallibleFunctionTransform>),
```

Comment:
I wonder if we don't have this backward. The actual impl for `Transform` is pretty bare, functions to create a `Transform` from wrapped types or destruct the `Transform` into its inner type and we're adding to that here. I wonder if we couldn't just drop the `Transform` enum entirely -- we already panic if the wrong transform is coerced in the wrong place so I don't think the variant wins us all that much -- or unify the interfaces of `FunctionTransform` and `FalliableFunctionTransform`? The difference so far is pretty small. 

When possible it'd be pretty nice to drop dynamic dispatch, or avoid adding to it. 

Without reference to whether its feasible to achieve in this PR or a related one or not, it'd be real great to convert this enum to a It might be too big a shift for